File: functions/api/lookup.js
/**
 * Cloudflare Pages Function: /api/lookup
 * 
 * Lightweight serverless quick-lookup endpoint for any stock ticker or company name
 * outside the curated watchlist.
 * Calls Finnhub's /quote, /company-news, and /search (for company name to ticker resolution) live
 * and returns price + 2-3 recent headlines.
 * 
 * Query Parameters:
 *   - ticker, symbol, q, or query: Stock symbol or company name (e.g. NFLX, Netflix, TSM, Spotify, PLTR)
 * 
 * Response Format:
 *   {
 *     "found": true,
 *     "symbol": "NFLX",
 *     "company_name": "NETFLIX INC",
 *     "resolved_from": "Netflix",
 *     "price": {
 *       "current": 690.50,
 *       "change": 8.20,
 *       "change_pct": 1.20,
 *       "high": 695.10,
 *       "low": 685.80,
 *       "open": 688.00,
 *       "prev_close": 682.30,
 *       "timestamp": 1726430400
 *     },
 *     "headlines": [
 *       {
 *         "headline": "...",
 *         "url": "https://...",
 *         "source": "Reuters",
 *         "datetime": 1726425600,
 *         "time_ago": "2h ago",
 *         "summary": "..."
 *       }
 *     ],
 *     "queried_at": "2026-09-17T15:30:00.000Z"
 *   }
 */

export async function onRequestGet(context) {
  return handleLookup(context);
}

export async function onRequest(context) {
  if (context.request.method === "OPTIONS") {
    return new Response(null, {
      status: 204,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type",
      },
    });
  }
  return handleLookup(context);
}

async function handleLookup(context) {
  const url = new URL(context.request.url);
  const rawQuery = url.searchParams.get("ticker") || 
                   url.searchParams.get("symbol") || 
                   url.searchParams.get("q") || 
                   url.searchParams.get("query") || "";
  const query = rawQuery.trim();

  // Validate query parameter
  if (!query) {
    return new Response(
      JSON.stringify({
        error: "Invalid or missing query parameter. Provide a stock symbol or company name (e.g. ?ticker=NFLX or ?q=Netflix).",
        symbol: null,
        found: false,
      }),
      {
        status: 400,
        headers: {
          "Content-Type": "application/json",
          "Access-Control-Allow-Origin": "*",
        },
      }
    );
  }

  // Retrieve Finnhub API Key from Cloudflare Environment Bindings
  const apiKey = context.env?.FINNHUB_API_KEY || "";
  if (!apiKey) {
    return new Response(
      JSON.stringify({
        error: "FINNHUB_API_KEY environment variable is not configured on Cloudflare Pages.",
        symbol: query.toUpperCase(),
        found: false,
      }),
      {
        status: 500,
        headers: {
          "Content-Type": "application/json",
          "Access-Control-Allow-Origin": "*",
        },
      }
    );
  }

  const fetchOptions = {
    headers: {
      "User-Agent": "StockNewsDashboard-Lookup/1.0",
      "Accept": "application/json",
    },
  };

  const isDirectTickerCandidate = /^[A-Z0-9.\-]{1,6}$/i.test(query);
  let resolvedTicker = query.toUpperCase();
  let resolvedCompanyName = null;
  let resolvedFrom = null;

  try {
    let quote = null;
    let news = [];
    let needSymbolSearch = !isDirectTickerCandidate;

    // Helper to fetch quote and news for a specific ticker symbol
    async function fetchQuoteAndNews(sym) {
      const now = new Date();
      const toDate = now.toISOString().split("T")[0];
      const fromDateObj = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
      const fromDate = fromDateObj.toISOString().split("T")[0];

      const quoteUrl = `https://finnhub.io/api/v1/quote?symbol=${encodeURIComponent(sym)}&token=${encodeURIComponent(apiKey)}`;
      const newsUrl = `https://finnhub.io/api/v1/company-news?symbol=${encodeURIComponent(sym)}&from=${fromDate}&to=${toDate}&token=${encodeURIComponent(apiKey)}`;

      const [qRes, nRes] = await Promise.all([
        fetch(quoteUrl, fetchOptions),
        fetch(newsUrl, fetchOptions),
      ]);

      if (qRes.status === 401 || nRes.status === 401) {
        throw new Error("UNAUTHORIZED");
      }

      let qData = null;
      if (qRes.ok) {
        qData = await qRes.json();
      }

      let nData = [];
      if (nRes.ok) {
        const rawNews = await nRes.json();
        if (Array.isArray(rawNews)) {
          nData = rawNews;
        }
      }

      return { quote: qData, news: nData };
    }

    // 1. If candidate ticker format, attempt direct quote check
    if (isDirectTickerCandidate) {
      const res = await fetchQuoteAndNews(resolvedTicker);
      quote = res.quote;
      news = res.news;

      const hasValidPrice = quote && (quote.c !== 0 || quote.pc !== 0 || quote.h !== 0);
      const hasNews = news && news.length > 0;

      if (!hasValidPrice && !hasNews) {
        // Direct symbol returned no data, fallback to symbol search
        needSymbolSearch = true;
      }
    }

    // 2. Resolve company name to ticker via Finnhub symbol search endpoint (/search?q=...)
    if (needSymbolSearch) {
      const searchUrl = `https://finnhub.io/api/v1/search?q=${encodeURIComponent(query)}&token=${encodeURIComponent(apiKey)}`;
      const searchRes = await fetch(searchUrl, fetchOptions);

      if (searchRes.status === 401) {
        throw new Error("UNAUTHORIZED");
      }

      if (searchRes.ok) {
        const searchData = await searchRes.json();
        const results = Array.isArray(searchData?.result) ? searchData.result : [];

        if (results.length > 0) {
          // Priority matching:
          // 1) Exact symbol match
          // 2) US Common Stock / primary listing (no exchange dots like .TO, .L)
          // 3) Any Common Stock or EQS type
          // 4) First result
          const best = results.find(r => r.symbol && r.symbol.toUpperCase() === query.toUpperCase()) ||
                       results.find(r => r.type === "Common Stock" && !r.symbol.includes(".")) ||
                       results.find(r => (r.type === "Common Stock" || r.type === "EQS") && r.symbol) ||
                       results[0];

          if (best && best.symbol) {
            resolvedTicker = best.symbol.toUpperCase();
            resolvedCompanyName = best.description || null;
            resolvedFrom = query;

            const res = await fetchQuoteAndNews(resolvedTicker);
            quote = res.quote;
            news = res.news;
          }
        }
      }
    }

    // Finnhub returns all zeros for nonexistent/invalid stock symbols
    const hasValidPrice = quote && (quote.c !== 0 || quote.pc !== 0 || quote.h !== 0);
    const hasNews = news && news.length > 0;

    if (!hasValidPrice && !hasNews) {
      return new Response(
        JSON.stringify({
          found: false,
          symbol: resolvedTicker || query.toUpperCase(),
          resolved_from: resolvedFrom,
          message: `No quote or news found on Finnhub for "${query}".`,
        }),
        {
          status: 404,
          headers: {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Cache-Control": "public, max-age=60",
          },
        }
      );
    }

    // Extract top 2-3 most recent headlines
    const topHeadlines = (news || [])
      .filter((item) => item && item.headline && item.url)
      .slice(0, 3)
      .map((item) => {
        let timeAgo = "";
        if (item.datetime) {
          const diffSec = Math.floor(Date.now() / 1000 - item.datetime);
          if (diffSec < 60) timeAgo = "Just now";
          else if (diffSec < 3600) timeAgo = `${Math.floor(diffSec / 60)}m ago`;
          else if (diffSec < 86400) timeAgo = `${Math.floor(diffSec / 3600)}h ago`;
          else timeAgo = `${Math.floor(diffSec / 86400)}d ago`;
        }
        return {
          headline: item.headline.trim(),
          url: item.url,
          source: item.source || "Financial Press",
          datetime: item.datetime || null,
          time_ago: timeAgo,
          summary: item.summary ? item.summary.trim() : "",
        };
      });

    const responsePayload = {
      found: true,
      symbol: resolvedTicker,
      company_name: resolvedCompanyName,
      resolved_from: resolvedFrom,
      price: hasValidPrice
        ? {
            current: Number(quote.c) || 0,
            change: Number(quote.d) || 0,
            change_pct: Number(quote.dp) || 0,
            high: Number(quote.h) || 0,
            low: Number(quote.l) || 0,
            open: Number(quote.o) || 0,
            prev_close: Number(quote.pc) || 0,
            timestamp: quote.t || Math.floor(Date.now() / 1000),
          }
        : null,
      headlines: topHeadlines,
      queried_at: new Date().toISOString(),
    };

    return new Response(JSON.stringify(responsePayload), {
      status: 200,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Cache-Control": "public, max-age=60, s-maxage=120",
      },
    });
  } catch (err) {
    if (err.message === "UNAUTHORIZED") {
      return new Response(
        JSON.stringify({
          error: "Unauthorized: Invalid FINNHUB_API_KEY configured.",
          symbol: resolvedTicker,
          found: false,
        }),
        {
          status: 401,
          headers: {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
          },
        }
      );
    }

    return new Response(
      JSON.stringify({
        error: "Failed to fetch live data from Finnhub",
        details: err.message,
        symbol: resolvedTicker,
        found: false,
      }),
      {
        status: 502,
        headers: {
          "Content-Type": "application/json",
          "Access-Control-Allow-Origin": "*",
        },
      }
    );
  }
}

File: pipeline/lookup.py
"""Lightweight Quick-Lookup Engine for Non-Watchlist Tickers.

Directly queries Finnhub's /quote, /company-news, and /search endpoints and parses
pricing + recent headlines without heavyweight scoring, classification, or AI processing.
Supports both stock tickers (e.g. NFLX, TSM) and company names (e.g. Netflix, Spotify).
"""

from datetime import datetime, timedelta
import json
import logging
import os
import re
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, List, Optional

logger = logging.getLogger("quick_lookup")

FINNHUB_QUOTE_URL = "https://finnhub.io/api/v1/quote"
FINNHUB_NEWS_URL = "https://finnhub.io/api/v1/company-news"
FINNHUB_SEARCH_URL = "https://finnhub.io/api/v1/search"


def format_time_ago(unix_timestamp: Optional[int]) -> str:
    """Format Unix timestamp into human-readable relative time (e.g. '2h ago')."""
    if not unix_timestamp:
        return ""
    diff_sec = int(datetime.utcnow().timestamp() - unix_timestamp)
    if diff_sec < 60:
        return "Just now"
    if diff_sec < 3600:
        return f"{diff_sec // 60}m ago"
    if diff_sec < 86400:
        return f"{diff_sec // 3600}h ago"
    return f"{diff_sec // 86400}d ago"


def fetch_ticker_quick_lookup(
    ticker: str,
    api_key: Optional[str] = None,
    timeout: int = 8,
) -> Dict[str, Any]:
    """Fetch live quote and 2-3 headlines for any ticker or company name from Finnhub."""
    raw_query = (ticker or "").strip()
    if not raw_query:
        return {
            "found": False,
            "symbol": raw_query,
            "error": "Missing ticker parameter",
        }

    if api_key is not None:
        token = api_key.strip()
    else:
        token = os.environ.get("FINNHUB_API_KEY", "").strip()
        if not token:
            env_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), ".env")
            if os.path.exists(env_path):
                with open(env_path, "r", encoding="utf-8") as f:
                    for line in f:
                        line = line.strip()
                        if line and not line.startswith("#") and "FINNHUB_API_KEY=" in line:
                            token = line.split("=", 1)[1].strip().strip("'\"")
                            break

    if not token:
        logger.warning("FINNHUB_API_KEY not configured for quick-lookup.")
        return {
            "found": False,
            "symbol": raw_query.upper(),
            "error": "FINNHUB_API_KEY environment variable not configured",
        }

    is_direct_ticker_candidate = bool(re.match(r"^[A-Za-z0-9.\-]{1,6}$", raw_query))
    resolved_symbol = raw_query.upper()
    resolved_company_name = None
    resolved_from = None
    need_symbol_search = not is_direct_ticker_candidate

    def _fetch_quote_and_news(sym: str) -> tuple:
        q_data = {}
        n_items = []
        # Fetch Quote
        q_url = f"{FINNHUB_QUOTE_URL}?symbol={urllib.parse.quote(sym)}&token={urllib.parse.quote(token)}"
        try:
            req = urllib.request.Request(
                q_url,
                headers={"User-Agent": "StockNewsDashboard-Lookup/1.0", "Accept": "application/json"},
            )
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                q_data = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            if e.code == 401:
                raise PermissionError("Unauthorized: Invalid FINNHUB_API_KEY")
            logger.warning("Finnhub quote HTTP error for %s: %s", sym, e)
        except Exception as e:
            logger.warning("Finnhub quote network error for %s: %s", sym, e)

        # Fetch News
        now_dt = datetime.utcnow()
        to_date_str = now_dt.strftime("%Y-%m-%d")
        from_date_str = (now_dt - timedelta(days=7)).strftime("%Y-%m-%d")
        n_url = f"{FINNHUB_NEWS_URL}?symbol={urllib.parse.quote(sym)}&from={from_date_str}&to={to_date_str}&token={urllib.parse.quote(token)}"
        try:
            req = urllib.request.Request(
                n_url,
                headers={"User-Agent": "StockNewsDashboard-Lookup/1.0", "Accept": "application/json"},
            )
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                raw_n = json.loads(resp.read().decode("utf-8"))
                if isinstance(raw_n, list):
                    n_items = raw_n
        except urllib.error.HTTPError as e:
            if e.code == 401:
                raise PermissionError("Unauthorized: Invalid FINNHUB_API_KEY")
            logger.warning("Finnhub news HTTP error for %s: %s", sym, e)
        except Exception as e:
            logger.warning("Finnhub news network error for %s: %s", sym, e)

        return q_data, n_items

    quote_data: Dict[str, Any] = {}
    news_items: List[Dict[str, Any]] = []

    try:
        if is_direct_ticker_candidate:
            quote_data, news_items = _fetch_quote_and_news(resolved_symbol)
            has_valid_price = bool(
                quote_data and (quote_data.get("c") or quote_data.get("pc") or quote_data.get("h"))
            )
            has_news = bool(news_items)
            if not has_valid_price and not has_news:
                need_symbol_search = True

        if need_symbol_search:
            s_url = f"{FINNHUB_SEARCH_URL}?q={urllib.parse.quote(raw_query)}&token={urllib.parse.quote(token)}"
            try:
                req = urllib.request.Request(
                    s_url,
                    headers={"User-Agent": "StockNewsDashboard-Lookup/1.0", "Accept": "application/json"},
                )
                with urllib.request.urlopen(req, timeout=timeout) as resp:
                    s_data = json.loads(resp.read().decode("utf-8"))
                    results = s_data.get("result", []) if isinstance(s_data, dict) else []
                    if results:
                        # Priority matching
                        best = None
                        for r in results:
                            sym = r.get("symbol", "")
                            if sym.upper() == raw_query.upper():
                                best = r
                                break
                        if not best:
                            for r in results:
                                if r.get("type") == "Common Stock" and "." not in r.get("symbol", ""):
                                    best = r
                                    break
                        if not best:
                            for r in results:
                                if r.get("type") in ("Common Stock", "EQS") and r.get("symbol"):
                                    best = r
                                    break
                        if not best:
                            best = results[0]

                        if best and best.get("symbol"):
                            resolved_symbol = best["symbol"].upper()
                            resolved_company_name = best.get("description")
                            resolved_from = raw_query
                            quote_data, news_items = _fetch_quote_and_news(resolved_symbol)
            except urllib.error.HTTPError as e:
                if e.code == 401:
                    return {"found": False, "symbol": resolved_symbol, "error": "Unauthorized: Invalid FINNHUB_API_KEY"}
                logger.warning("Finnhub search HTTP error for %s: %s", raw_query, e)
            except Exception as e:
                logger.warning("Finnhub search network error for %s: %s", raw_query, e)

    except PermissionError as e:
        return {"found": False, "symbol": resolved_symbol, "error": str(e)}

    # Validate if valid price or news exists
    has_valid_price = bool(
        quote_data and (quote_data.get("c") or quote_data.get("pc") or quote_data.get("h"))
    )
    has_news = bool(news_items)

    if not has_valid_price and not has_news:
        return {
            "found": False,
            "symbol": resolved_symbol or raw_query.upper(),
            "resolved_from": resolved_from,
            "message": f'No quote or news found on Finnhub for "{raw_query}".',
        }

    # Format 2-3 recent headlines
    top_headlines = []
    for item in news_items[:3]:
        if not isinstance(item, dict):
            continue
        headline = (item.get("headline") or "").strip()
        url = item.get("url") or ""
        if headline and url:
            dt = item.get("datetime")
            top_headlines.append({
                "headline": headline,
                "url": url,
                "source": item.get("source") or "Financial Press",
                "datetime": dt,
                "time_ago": format_time_ago(dt),
                "summary": (item.get("summary") or "").strip(),
            })

    now = datetime.utcnow()
    price_dict = None
    if has_valid_price:
        price_dict = {
            "current": float(quote_data.get("c", 0.0) or 0.0),
            "change": float(quote_data.get("d", 0.0) or 0.0),
            "change_pct": float(quote_data.get("dp", 0.0) or 0.0),
            "high": float(quote_data.get("h", 0.0) or 0.0),
            "low": float(quote_data.get("l", 0.0) or 0.0),
            "open": float(quote_data.get("o", 0.0) or 0.0),
            "prev_close": float(quote_data.get("pc", 0.0) or 0.0),
            "timestamp": int(quote_data.get("t") or int(now.timestamp())),
        }

    return {
        "found": True,
        "symbol": resolved_symbol,
        "company_name": resolved_company_name,
        "resolved_from": resolved_from,
        "price": price_dict,
        "headlines": top_headlines,
        "queried_at": now.isoformat() + "Z",
    }
